src/Credentials/InstanceProfileProvider.php

```diff
@@ -296,4 +305,29 @@ private function decodeResult($response)
 
         return $result;
     }
+
+    /**
+     * This functions checks for whether we should fall back to IMDSv1 or not.
+     * If $ec2MetadataV1Disabled is null then we will try to resolve this value from
+     * the following sources:
+     * - From environment: "AWS_EC2_METADATA_V1_DISABLED".
+     * - From config file: aws_ec2_metadata_v1_disabled
+     * - Defaulted to false
+     *
+     * @return bool
+     */
+    private function shouldFallbackToIMDSv1() {
+        $isImdsV1Disabled = boolean_value($this->ec2MetadataV1Disabled)
+            ?? boolean_value(
```

Comment:
Also, let's add a return type to this.

Reply:
>I don't think we need the boolean_value conversion here.

Yeah, I know what you mean, but I am trying to avoid a false positive here. If any string is found, regardless if is true or false, in any of the source from where we try to resolve the value then, the implementation will interpret that resolved value as true and it will return false, which means it will not fallback to IMDSv1. Please see below:

*Lets say $ec2MetadataV1Disabled is set to "dummyvalue"*
```php
private function shouldFallbackToIMDSv1(): bool {
    $isImdsV1Disabled = $this->ec2MetadataV1Disabled
        ?? ConfigurationResolver::resolve(
                self::CFG_EC2_METADATA_V1_DISABLED,
                self::DEFAULT_AWS_EC2_METADATA_V1_DISABLED,
                'bool',
                ['use_aws_shared_config_files' => true]
            )
        ?? self::DEFAULT_AWS_EC2_METADATA_V1_DISABLED;

    return !$isImdsV1Disabled;
}
```
In the return statement $isImdsV1Disabled will be equals to "dummyvalue", which in the boolean expression it will be interpreted as truly value, and therefore a false will be returned due to that we are turning the negation of the resolved value, which in this case will be the negation true that is false. However, when using boolean_value, if the value is not a boolean a null will be returned. Should we maybe consider throw an error when the value resolved is not the expected type?

What do you think?

Reply:
I don't think we need to worry about that because of this: https://github.com/aws/aws-sdk-php/blob/41c7a9491455f89052cba62c4e4f0c748af51265/src/Configuration/ConfigurationResolver.php#L195C8-L199C10